[Windows] Bug fixes for MSVC

willyd opened the pull request:
- file_store_handler.cc: `mkdir` only accepts one argument and requires inclusion of <direct.h>
- math.h: macro workaround does not work for integerIsPowerOf2 when prefixed with `math` namespace.
- GpuBitonicSort.cuh: use `std::integral_constant` since nvcc ignores constexpr with MSVC (fixes #997)
- pool_op_cudnn.cu: undefine `IN` and `OUT` macros defined in `minwindef.h`
- logging.cc: Prefix glog logging levels with name since MSVC cannot use the abbreviated macros

Yangqing commented:
Thanks! We had some glitches in the pull request import process, and it's solved now. Will merge.

Yangqing commented on caffe2/utils/GpuBitonicSort.cuh:
```diff
@@ -39,9 +39,11 @@ __device__ inline void bitonicSort(K* keys,
   //               "sort data too large (>32768 bytes)");
-  static_assert(math::integerIsPowerOf2(Power2SortSize),
+  static_assert(math::integerIsPowerOf2(
+                std::integral_constant<int, Power2SortSize>()),
```
Yep, this seems to be needed, otherwise build errors.

Yangqing commented on caffe2/distributed/file_store_handler.cc:
```diff
@@ -6,6 +6,9 @@
 #include <stdlib.h>
 #include <sys/stat.h>
+#if defined(_MSC_VER)
+#include <direct.h>
```
Kindly rebase and remove this?

Yangqing commented on caffe2/utils/GpuBitonicSort.cuh:
```diff
@@ -39,9 +39,11 @@ __device__ inline void bitonicSort(K* keys,
                 "sort size must be power of 2");
-  static_assert(math::integerIsPowerOf2(ThreadsPerBlock),
+  static_assert(math::integerIsPowerOf2(
+                std::integral_constant<int, ThreadsPerBlock>()),
```
Here as well

Yangqing commented on caffe2/utils/GpuBitonicSort.cuh:
```diff
@@ -107,7 +109,8 @@ __device__ inline void warpBitonicSort(K* keys,
                 "sort not large enough");
-  static_assert(math::integerIsPowerOf2(Power2SortSize),
+  static_assert(math::integerIsPowerOf2(
+                std::integral_constant<int, Power2SortSize>()),
```
ditto

willyd commented:
@Yangqing Thanks for the review. I rebased and removed the requested #include. Plus I fixed a few other issues. One that you might want to look into is the missing CMakeLists.txt file in `contrib/opengl`.

Yangqing commented:
Thanks @willyd - yeah, we traced the issue back to our sync script ignoring the CMakeLists.txt files during syncing, and we are fixing it. Let me pull the PR and land it if all checks out.

Yangqing commented:
@willyd do you mind rebase to master, and also address the std::integral_constant<int, Power2SortSize>()::value comment?

willyd commented:
@Yangqing Rebase done. Also addressed the `::value` comment.

harouwu commented:
@Yangqing @willyd  what is the update on this PR? If it looks good, @willyd would you rebase and we are ready to go?

willyd commented:
@harouwu I rebased on master last time @Yangqing asked. Since then 69 new commits have been pushed to master, but none of them have conflicting changes as per github. I don't see the need for another rebase, can you please elaborate on why I would need to rebase again?

I am basically waiting for his feedback or for him to merge the PR.

Callum17 commented:
Did a fresh build from latest master commit @ 799d7ccaae3598c9e04712195632028423e8f83c. Merged pull this request @ 67849563d69a78b262d2c61676a7c6b1c8a92dc9. Successful compilation with the following change:

Minor issue where MSVC complains about type narrowing conversions in {base}\caffe2\caffe2\layer_norm_op.cu at lines 264 and 294.

264: gscratch_.Resize(std::vector<size_t>{left, right});
should be
264: gscratch_.Resize(std::vector<size_t>{static_cast<size_t>(left), static_cast<size_t>(right)});

294: dstdev_.Resize(vector<size_t>{left, 1});
should be
294: dstdev_.Resize(vector<size_t>{static_cast<size_t>(left), 1});

Looks promising after successful convnet_benchmark runs for AlexNet, OverFeat, Inception. 

VGGA model crashes. MLP returns with error:

Traceback (most recent call last):
  File "D:\python27\Lib\runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "D:\python27\Lib\runpy.py", line 72, in _run_code
    exec code in run_globals
  File "C:\Users\Callum\Documents\GitHub\caffe2\build\caffe2\experiments\python\convnet_benchmarks.py", line 685, in <module>
    Benchmark(model_map[args.model], args)
  File "C:\Users\Callum\Documents\GitHub\caffe2\build\caffe2\experiments\python\convnet_benchmarks.py", line 573, in Benchmark
    AddParameterUpdate(model)
  File "C:\Users\Callum\Documents\GitHub\caffe2\build\caffe2\experiments\python\convnet_benchmarks.py", line 534, in AddParameterUpdate
    param_grad = model.param_to_grad[param]
KeyError: BlobReference("fc_1_0_w")

willyd commented:
Gentle reminder for @Yangqing.

I just updated the PR to latest master and fixed the issues mentioned by @Callum17 and some others. Should I squash the commits into a single one?

pengfei666 commented:
@willyd This PR is very useful, I have compiled successfully, thanks. Hope it could be merged soon.

willyd commented:
I just added a few more commits to support building with a dynamic runtime library and to support building with vcpkg. I also submitted a [PR](https://github.com/Microsoft/vcpkg/pull/1816) to vcpkg. If anyone is interested in testing it, it would be much appreciated.

Commit a7ece7c3611e0cf2736a7c0ce121b83dfc0d0d29 should fix #403.

willyd commented:
caffe2 (CPU and C++ only) is now available on windows via [vcpkg](https://github.com/Microsoft/vcpkg). Installing should be as simple as: `vcpkg install caffe2 --triplet x64-windows` or `vcpkg install caffe2 --triplet x64-windows-static`.

hughsando commented:
Not sure if this is the best place to mention this - and it might be covered by the PRs here - some of the other issues I had were.
In caffe2/perfkernels/embedding_lookup_avx2.cc, the CAFFE2_ALIGNED(64) macro should come before the variable declarations, and _mm_prefetch needs a type-cast for the point arg in msvc.  This is auto-generated so the fixes will be elsewhere.

aralph commented on CMakeLists.txt:
```diff
@@ -56,6 +56,14 @@ option(USE_ROCKSDB "Use RocksDB" ON)
 option(USE_THREADS "Use Threads" ON)
 option(USE_ZMQ "Use ZMQ" OFF)
+if(MSVC)
+  if(BUILD_BUILD_SHARED_LIBS)
```
Shouldn't the variable be 'BUILD_SHARED_LIBS'?

willyd replied:
@aralph You are right! Thanks for pointing out. I must'nt have had enough coffee when I wrote this. When I get back to caffe2 on Windows I will update the PR.

jimmyoic commented:
Hi,
Did anyone else get the problems about link error?
I follow the post (https://research.wmz.ninja/articles/2017/05/build-caffe2-on-windows-10-gpu.html), but I kept getting the error message ( undefined external symbol...)

I tried add the dependencies of cuda library in VS 2015, but it didn't work.

orionr commented:
Thank you for your contribution! PyTorch and Caffe2 are now officially merged with the PyTorch 1.0 preview release. Please submit on https://github.com/pytorch/pytorch if this is still an issue over there.
